### PR TITLE
Implement read package translations from database

### PR DESCRIPTION
**Pull request by GabrielEValenzuela**
|Related issue|
|---|
|#19044|

## Objective

This pull request introduces the implementation of a significant enhancement to our Vulnerability Scanner, focusing on the introduction of a translation package mechanism for Windows systems along with a cache mechanism using the Least Recently Used (LRU) policy. This enhancement aims to improve the performance of reading package translations from the database, thereby reducing database load and latency for end-users.

## Note about cache ($) algorithm
<details><summary>Expand</summary>

In computing, cache replacement policies (also frequently called cache replacement algorithms or cache algorithms) are optimizing instructions or algorithms, that a computer program or a hardware-maintained structure can utilize in order to manage a cache of information stored on the computer. Caching improves performance by keeping recent or often-used data items in memory locations that are faster or computationally cheaper to access than normal memory stores. When the cache is full, the algorithm must choose which items to discard to make room for the new ones.

Traditional caching policies such as ﬁrst in ﬁrst out (FIFO), least recently used (LRU), least frequently used (LFU) and their variants have been widely applied in several applications, where there are abundant caching and computing resources and the served area is usually very large. The LRU policy replaces the cache entry that has not been used for the longest time in the past and it's very useful for programs that exhibit spatial and temporal locality.

When it comes to LRU cache analysis, the state-of-the-art currently presents a tradeoff between precision and analysis efficiency: The classical approach to the static analysis of LRU caches [Ferdinand and Wilhelm 1999] is a highly-efficient abstract interpretation that essentially keeps for each block a range of how many other blocks have been used more recently. 

In summary, The “Least Recently Used” (LRU) replacement policy is simple and intuitive: the data block least recently used is evicted when a cache miss occurs. The cache is thus a queue ordered by age: on a miss, the oldest block is discarded to make room for a new one, which has age 0; the ages of all other blocks are incremented. If a block is already in the cache, it is “rejuvenated”: its age is set to zero, and the ages of the blocks before it in the queue are incremented. In other words, the state of an LRU cache with associativity N is a word of length at most N over the alphabet of cache blocks, composed of pairwise distinct letters; an empty cache is defined by the empty word. When an access is made to a block a, if it belongs to the word (hit), then this letter is removed from the word and appended to the word. If it does not belong to the word (miss), and the length of the word is less than N, then a is appended to the word; otherwise, the length of the word is exactly N — the first letter of the word is discarded and a is appended.

Ref:
- David Monniaux, Valentin Touzeau. On the complexity of cache analysis for different replacement
policies. Journal of the ACM (JACM), 2019, 66 (6), pp.41. ff10.1145/3366018ff. ffhal-01910216v3
- Valentin Touzeau et al. “Fast and exact analysis for LRU caches”. In: Proceedings of the ACM on Programming Languages (PACMPL) 3.POPL (2019), 54:1–54:29. doi: 10.1145/3290367. arXiv: 1811.01670.

</details>

## Description
- Implemented validation logic to verify the content of the package translation object.
- Added error-handling mechanisms to handle invalid or incomplete configurations.
- Included unit tests to verify the accuracy of the validation process.

For a better compression of the detailed work, please check this top-level sequence diagram

<details><summary>Sequence Diagram</summary>

image

</details>

## Quality Assurance
A new test has been included to verify the correct behavior and backward compatibility of this solution proposal.

<details><summary>Output testing 🟢 </summary>

```bash
[==========] Running 5 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 5 tests from DatabaseFeedManagerTest
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationCacheMiss
Cache miss for package 'Apache Tomcat.GenericVersion'
[       OK ] DatabaseFeedManagerTest.PackageTranslationCacheMiss (193 ms)
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationCacheHit
Cache miss for package 'Apache Tomcat.GenericVersion'
Cache hit for package 'Apache Tomcat.GenericVersion'
[       OK ] DatabaseFeedManagerTest.PackageTranslationCacheHit (154 ms)
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationNoTranslationFound
Cache miss for package 'SQLite3'
No translation available for package 'SQLite3'
[       OK ] DatabaseFeedManagerTest.PackageTranslationNoTranslationFound (147 ms)
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationInvalidDatabase
Cache miss for package 'Apache Tomcat.GenericVersion'
[       OK ] DatabaseFeedManagerTest.PackageTranslationInvalidDatabase (145 ms)
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationInvalidCache
[       OK ] DatabaseFeedManagerTest.PackageTranslationInvalidCache (33 ms)
[----------] 5 tests from DatabaseFeedManagerTest (674 ms total)

[----------] Global test environment tear-down
[==========] 5 tests from 1 test suite ran. (674 ms total)
[  PASSED  ] 5 tests.


[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from CacheLRUTest
[ RUN      ] CacheLRUTest.insertAndHit
[       OK ] CacheLRUTest.insertAndHit (0 ms)
[ RUN      ] CacheLRUTest.insertAndMiss
[       OK ] CacheLRUTest.insertAndMiss (0 ms)
[----------] 2 tests from CacheLRUTest (0 ms total)
[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (1 ms total)
[  PASSED  ] 2 tests.

```

</details>

Regarding memory, no new memory allocations were introduced in this PR using `raw_pointers`.

<details><summary>ScanBuild report</summary>

Command:

```bash
scan-build-15 -enable-checker alpha.core.CastSize -enable-checker alpha.core.CastToStruct -enable-checker alpha.core.Conversion -enable-checker alpha.core.IdenticalExpr -enable-checker alpha.core.SizeofPtr -enable-checker alpha.core.TestAfterDivZero --exclude external make TARGET=server DEBUG=1 -j$(nproc)
```

```bash

Done building server

scan-build: No bugs found.

```

</details>

<details><summary>Clang-Format</summary>

```bash
---------------------------------
PASSED: Clang-format check passed
---------------------------------
```
</details>

## Exploratory tests
Not apply

### Check summary
- [x] Style and quality of SCA.
- [x] Documentation clear.
- [x] Functionality optimizated.

### CODE REVIEW

File: src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityTranslationPackages.fbs
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/schemas/vulnerabilityTranslationPackages.fbs:1]: This file shouldn't be the same as the one from branch issue #19047 of @MiguelazoDS?

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp [hunk 1]
```diff
@@ -158,6 +166,83 @@ class TDatabaseFeedManager final : public Observer<nlohmann::json&>
...
+     *       and querying package translations.
+     */
+    void getWazuhPackageTranslation(const std::string& packageName,
+                                    const std::function<void(const NSVulnerabilityScanner::srcTrlData&)>& callback)
+    {
+        auto resultCache = m_translationsCache->getValue(packageName);
+
+        if (resultCache.has_value())
+        {
+            std::cerr << "Cache hit for package '" << packageName << "'\n";
```
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:191]: Is the log to stderr correct? Or should be to stdout?
Is this a possible normal situation to be logging to stderr?

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp [hunk 2]
```diff
@@ -158,6 +166,83 @@ class TDatabaseFeedManager final : public Observer<nlohmann::json&>
...
+            auto translations = dtoVulnTranslation.data->data()->translations();
+            for (unsigned int index = 0; index < translations->size(); index++)
+            {
+                callback(*(translations->Get(index)));
+            }
+
+            return;
+        }
+
+        std::cerr << "Cache miss for package '" << packageName << "'\n";
```
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:220]: Is the log to stderr correct? Or should be to stdout?
Is this a possible normal situation to be logging to stderr?

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp (hunk 1 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:191]: The only way to see an output during the tests is with the `stderr`, later this message will be put in a login function

File: src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp (hunk 2 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/src/databaseFeedManager/databaseFeedManager.hpp:220]: Ditto

File: src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp [hunk 3]
```diff
@@ -537,3 +537,224 @@ TEST_F(DatabaseFeedManagerTest, GetVulnerabilityRemediation_InvalidData)
...
+    // Call function
+    spDatabaseFeedManager->getWazuhPackageTranslation(
+        packageName,
+        [&translations](const NSVulnerabilityScanner::srcTrlData& translation)
+        { translations.push_back(translation.product()->str()); });
+
+    // Assert
+    EXPECT_EQ(translations.size(), 2);
+    EXPECT_EQ(translations.at(0), "tomcat");
+    EXPECT_EQ(translations.at(1), "tomcat");
```
Comment (matias-braida) [inline src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp:656]: The fact that the translation is found twice doesn't assure that it has got a cache hit

**Review by matias-braida — COMMENTED** [review]
Great Job!

File: src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp (hunk 3 above)
Comment (GabrielEValenzuela) [reply src/wazuh_modules/vulnerability_scanner/tests/databaseFeedManager/databaseFeedManager_test.cpp:656]: According to what we talked about earlier today, yes we tested two occurrences of the package and saw the correct behavior in an indirect way, with the test output we can see:

```bash
[ RUN      ] DatabaseFeedManagerTest.PackageTranslationCacheHit
Cache miss for package 'Apache Tomcat.GenericVersion'
Cache hit for package 'Apache Tomcat.GenericVersion'
[       OK ] DatabaseFeedManagerTest.PackageTranslationCacheHit (154 ms)
```

We can change the function signature, we can check with @Dwordcito to return a code `(-1,0,1)` for missing, database and cache scenario.

**Review by matias-braida — APPROVED** [review]
LGTM

**Pull request merged**